src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs

```diff
@@ -414,24 +416,44 @@ private void EmitObject(string objectFilePath, IReadOnlyCollection<DependencyNod
 
                 sectionWriter.EmitAlignment(nodeContents.Alignment);
 
+                bool haveStartSymbol = false;
                 bool isMethod = node is IMethodBodyNode or AssemblyStubNode;
                 long thumbBit = _nodeFactory.Target.Architecture == TargetArchitecture.ARM && isMethod ? 1 : 0;
                 foreach (ISymbolDefinitionNode n in nodeContents.DefinedSymbols)
                 {
+                    haveStartSymbol |= n.Offset == 0;
                     sectionWriter.EmitSymbolDefinition(
                         n == node ? currentSymbolName : GetMangledName(n),
                         n.Offset + thumbBit,
-                        n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0);
+                        n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0,
+                        altEntry: _usesSubsectionsViaSymbols && n.Offset != 0);
                     if (_nodeFactory.GetSymbolAlternateName(n) is string alternateName)
                     {
                         sectionWriter.EmitSymbolDefinition(
                             ExternCName(alternateName),
                             n.Offset + thumbBit,
                             n.Offset == 0 && isMethod ? nodeContents.Data.Length : 0,
-                            global: true);
+                            global: true,
+                            altEntry: _usesSubsectionsViaSymbols);
                     }
                 }
 
+                // In the subsections-via-symbols model (Mach-O file format used by Apple) we need a symbol pointing
+                // to the beginning on the node without the `altEntry` flag. All symbols with non-zero offset are then
+                // produced with the `altEntry: true` flag. This makes the node appear as unbreakable atom for the
+                // native linker with individual symbol labels pointing into it.
+                //
+                // If the defined symbols of the node don't cover the zero offset we have to synthesize the initial
+                // label.
+                if (!haveStartSymbol && _usesSubsectionsViaSymbols)
+                {
+                    string startSymbol = currentSymbolName != null ? $"__start_{currentSymbolName}" :$"__start_{syntheticLabelCounter++}";
```

Comment:
I'd appreciate suggestions for the name here. I just chose something that didn't collide with existing symbols but I am not particularly happy about the name. I could try to emit emit it as local symbol too but that will require a bit more tweaking. Currently it's "private extern" (aka "hidden" in ELF terminology).